### ISSUE TITLE
bug: fluentd unable to start with error="fork/exec /usr/local/bundle/bin/fluentd: no such file or directory"

### ISSUE THREAD

**Issue by joshuabaird** [fluent/fluent-operator#1187]
### Describe the issue

It seems a recent image was pushed to the `kubesphere/fluentd:1.15.3` tag (`docker.io/kubesphere/fluentd@sha256:bc06e880c224e76e659bf59250e5302ad159ee6b5474a2c5ee45f3a0969644c5`) which breaks fluentd:

```
level=error msg="start Fluentd error" error="fork/exec /usr/local/bundle/bin/fluentd: no such file or directory"
```

Pinning to a previous SHA fixes the issue -- `kubesphere/fluentd:v1.15.3@sha256:794311919658aee8eb9829836cd6c3437dffd9c7112556d5dc2f01ca3fcb826b`.

### To Reproduce

Repull the `kubesphere/fluentd:1.15.3` latest SHA.

### Expected behavior

Fluentd should start.

### Your Environment

```markdown
N/A
```

**Comment by joshuabaird** [comment]
@benjaminhuo @wenchajun Can someone please review this?

**Comment by m-gavrilyuk** [comment]
same error on fluent/fluent-operator/fluentd:v2.8.0

**Comment by rurus9** [comment]
I think that in general each new image should have a new tag (does not apply to floating tags, like "latest").

**Comment by joshuabaird** [comment]
> I think that in general each new image should have a new tag (does not apply to floating tags, like "latest").

I agree.  It's a little difficult here since the kubesphere image is tracking the upstream version of the

**Comment by joshuabaird** [comment]
I would agree.  Folks rely on versioned tags for stability and they should be immutable.  If these images are going to be rebuilt, perhaps an internal "patch" version should be added (eg, `v1.15.3.x`).

**Comment by benjaminhuo** [comment]
> I would agree. Folks rely on versioned tags for stability and they should be immutable. If these images are going to be rebuilt for whatever reason, perhaps an internal "patch" version should be added (eg, `v1.15.3.x`).

This might be related to the CI changes we made recently, cc @sarathchandra24

https://github.com/fluent/fluent-operator/pull/1183
https://github.com/fluent/fluent-operator/pull/1079

I also remember there is an PR for a similar issue from @sarathchandra24
https://github.com/fluent/fluent-operator/pull/1093

Would you help to take a look? @sarathchandra24

Thanks

**Comment by benjaminhuo** [comment]
I've built fluentd v1.17.0 image

![image](https://github.com/fluent/fluent-operator/assets/18525465/7e823191-b03f-4a5e-a3db-2a09ea67c4ac)

**Comment by joshuabaird** [comment]
@benjaminhuo It looks like the `1.17.0` image has the same bug.  Is this expected?

**Comment by sarathchandra24** [comment]
Sorry for the late response everyone, I realized the problem after running it locally.

Root cause is [defaultBinPath on main.go#L22](https://github.com/fluent/fluent-operator/blob/master/cmd/fluent-watcher/fluentd/main.go#L22) for amd64 it is"/usr/bin/fluentd" and for arm64 it is "/usr/local/bundle/bin/fluentd".

Creating a PR for logic to choose path based on arch.

**Comment by benjaminhuo** [comment]
> Sorry for the late response everyone, I realized the problem after running it locally.
> 
> Root cause is [defaultBinPath on main.go#L22](https://github.com/fluent/fluent-operator/blob/master/cmd/fluent-watcher/fluentd/main.go#L22) for amd64 it is"/usr/bin/fluentd" and for arm64 it is "/usr/local/bundle/bin/fluentd".
> 
> Creating a PR for logic to choose path based on arch.

Thank you very much @sarathchandra24

**Comment by benjaminhuo** [comment]
both 1.15.3 and 1.17.0 are updated, would you try again? @joshuabaird

**Comment by sarathchandra24** [comment]
@joshuabaird Can I know what OS are you using.

Also, I think there is something wrong with the builds or build system.

<img width="1074" alt="image" src="https://github.com/fluent/fluent-operator/assets/39090092/af90c391-dc43-4afa-a59f-bc522b20041d">

You see the message

```level=info msg="Current architecture" arch=amd64```


Also for 
``` docker run sarathchandra24/fluentd-arm:local-arm1```

<img width="756" alt="image" src="https://github.com/fluent/fluent-operator/assets/39090092/d6d7ea76-874f-4bda-9b6a-a5dfd12e98c9">

You see the message

```level=info msg="Current architecture" arch=arm64```

But this is not the case while running 
```docker run kubesphere/fluentd:1.17.0@sha256:bc06e880c224e76e659bf59250e5302ad159ee6b5474a2c5ee45f3a0969644c5```


<img width="1074" alt="image" src="https://github.com/fluent/fluent-operator/assets/39090092/dbbed234-3d73-4530-87f4-f86230a8c9db">

**Comment by sarathchandra24** [comment]
@joshuabaird Can you please run 

```docker run ghcr.io/fluent/fluent-operator/fluentd:v1.17@sha256:095572fbf94ee3bbd01c0597b7b8a113c647e64ad2c53457c9c561432207f99d```

and 

```docker run ghcr.io/fluent/fluent-operator/fluentd:v1.17@sha256:baac1724e2277baf50817d2612f06f0bf3b9050a77e1f7b78d351386b84541b7```

To check if GitHub images are working

After inspecting images on GitHub

running: ```docker run ghcr.io/fluent/fluent-operator/fluentd:v1.17@sha256:095572fbf94ee3bbd01c0597b7b8a113c647e64ad2c53457c9c561432207f99d```

<img width="1276" alt="image" src="https://github.com/fluent/fluent-operator/assets/39090092/c45631e2-f767-4e95-aa56-fe2bd7e33583">

We can see the message ```level=info msg="Current architecture" arch=amd64```

running: ```docker run ghcr.io/fluent/fluent-operator/fluentd:v1.17@sha256:baac1724e2277baf50817d2612f06f0bf3b9050a77e1f7b78d351386b84541b7```

<img width="1276" alt="image" src="https://github.com/fluent/fluent-operator/assets/39090092/ae07fa92-84e8-4a4e-ba8e-5324fc2e9dd3">

We can see the message ```level=info msg="Current architecture" arch=arm64```

**Comment by joshuabaird** [comment]
@benjaminhuo Also, just noticed that the fluentbit images aren't available in Github (ghcr.io) -- so we probably need to manually run the CI job that pushes them.

**Comment by joshuabaird** [comment]
It also looks like the `v1.17.0` `linux/amd64` image on GHCR is actually 1.15.3:

```
❯ docker run --platform linux/amd64 ghcr.io/fluent/fluent-operator/fluentd:v1.17.0                                                                                                      
level=info msg="Current architecture" arch=amd64
level=info msg="Fluentd started"

2024-06-05 16:03:02 +0000 [info]: init supervisor logger path=nil rotate_age=nil rotate_size=nil
2024-06-05 16:03:02 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/fluent.conf"
2024-06-05 16:03:02 +0000 [info]: gem 'fluentd' version '1.15.3'
...
2024-06-05 16:03:02 +0000 [info]: starting fluentd-1.15.3 pid=13 ruby="3.2.4"
2024-06-05 16:03:02 +0000 [info]: spawn command to main:  cmdline=["/usr/bin/ruby", "-Eascii-8bit:ascii-8bit", "/usr/bin/fluentd", "-c", "/fluentd/etc/fluent.conf", "-p", "/fluentd/plugins", "--under-supervisor"]
2024-06-05 16:03:02 +0000 [info]: init supervisor logger path=nil rotate_age=nil rotate_size=nil
```

The `linux/arm64` tag however is actually v1.17.0:

```
❯ docker run --platform linux/arm64 ghcr.io/fluent/fluent-operator/fluentd:v1.17.0
Unable to find image 'ghcr.io/fluent/fluent-operator/fluentd:v1.17.0' locally
v1.17.0: Pulling from fluent/fluent-operator/fluentd
Digest: sha256:4651f4340241b53534c5b481422082d9e785e4f9e86cd2d027a51f61e521fe2e
Status: Downloaded newer image for ghcr.io/fluent/fluent-operator/fluentd:v1.17.0
level=info msg="Current architecture" arch=arm64
level=info msg="Fluentd started"
2024-06-05 16:04:25 +0000 [info]: init supervisor logger path=nil rotate_age=nil rotate_size=nil
2024-06-05 16:04:25 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/fluent.conf"
2024-06-05 16:04:25 +0000 [info]: gem 'fluentd' version '1.17.0'
...
2024-06-05 16:04:25 +0000 [info]: starting fluentd-1.17.0 pid=14 ruby="3.3.2"
2024-06-05 16:04:25 +0000 [info]: spawn command to main:  cmdline=["/usr/local/bin/ruby", "-Eascii-8bit:ascii-8bit", "/usr/local/bundle/bin/fluentd", "-c", "/fluentd/etc/fluent.conf", "-p", "/fluentd/plugins", "--under-supervisor"]
2024-06-05 16:04:25 +0000 [info]: #0 init worker0 logger path=nil rotate_age=nil rotate_size=nil
2024-06-05 16:04:25 +0000 [info]: adding match in @FLUENT_LOG pattern="fluent.*" type="null"
2024-06-05 16:04:25 +0000 [info]: #0 starting fluentd worker pid=23 ppid=14 worker=0
2024-06-05 16:04:25 +0000 [info]: #0 fluentd worker is now running worker=0

**Comment by benjaminhuo** [comment]
@joshuabaird I've added you as the maintainer, and you can trigger the image build here:
![image](https://github.com/fluent/fluent-operator/assets/18525465/43c7a3b6-f493-4f58-950d-c07cb8fc8083)

**Comment by joshuabaird** [comment]
@benjaminhuo @sarathchandra24 Is the intention to build and maintain both v1.15.3 and v1.17.0 fluentd images?  Even if you pass `1.17.0` to the workflow, the `Dockerfile` still installs v1.15.3 here:

https://github.com/fluent/fluent-operator/blob/2de2b934e4d9f4475e278da7d9db74d89ef9a037/cmd/fluent-watcher/fluentd/Dockerfile.amd64#L30

So, if the intention is to build/maintain both v1.15.3 and 1.17.0, the `Dockerfile` will need to be modified.

**Comment by benjaminhuo** [comment]
> @benjaminhuo But do we want to continue to support v1.15.3 or just modify the Dockerfiles to use 1.17.0?

we already have 1.51.3 image built that meets some people's requirement, I think we can move on to the latest version of fluentd

**Comment by joshuabaird** [comment]
@benjaminhuo https://github.com/fluent/fluent-operator/pull/1198

**Comment by benjaminhuo** [comment]
> @benjaminhuo #1198

@joshuabaird The new fluentd image for 1.17 has been rebuilt after your PR, would you give it a try?
![image](https://github.com/fluent/fluent-operator/assets/18525465/cf10ca03-b483-4857-8502-53781d767f41)

**Comment by joshuabaird** [comment]
Things are looking good.  I'm going to open a PR to update fluentbit and then we'll rebuild the fluentbit images so they get pushed to GHCR.

**Comment by joshuabaird** [comment]
@benjaminhuo Any idea why a `fluentd:v2.8.0` and `fluent-bit:v2.8.0` image exists?

This is confusing, because it's the operator tag, not the fluentd/fluent-bit tag.  This is causing dependency update apps (like Dependabot/Renovate) to try and update these images.

Should we delete them?

<img width="1211" alt="image" src="https://github.com/fluent/fluent-operator/assets/6271918/b41956ad-072d-4648-a924-a6ad3898fdf6">

**Comment by benjaminhuo** [comment]
> @benjaminhuo Any idea why `fluentd:v2.8.0` and `fluent-bit:v2.8.0` tags exist?
> 
> This is confusing, because it's the operator tag, not the fluentd/fluent-bit tag. This is causing dependency update apps (like Dependabot/Renovate) to try and update these images.
> 
> Should we delete them?
> 
> <img alt="image" width="1211" src="https://private-user-images.githubusercontent.com/6271918/340787413-b41956ad-072d-4648-a924-a6ad3898fdf6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTg3NjM5ODMsIm5iZiI6MTcxODc2MzY4MywicGF0aCI6Ii82MjcxOTE4LzM0MDc4NzQxMy1iNDE5NTZhZC0wNzJkLTQ2NDgtYTkyNC1hNmFkMzg5OGZkZjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MTlUMDIyMTIzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzVkNjA1ZmY1ZTZkNTcxMzhhMTcyODMyMmFkN2QyMjc5ZTNjN2IyMDYzM2QyNGQwNGIxNzdkYWE2ZmEwMDZiMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Uyv5pkapM6_D_aJKmi4NAID0TqWKOTcrF_Nh6m_UVrY">

@joshuabaird I can delete them, they're created by wrong CI workflow

**Comment by benjaminhuo** [comment]
image 2.8.0 are all deleted

**Comment by joshuabaird** [comment]
@benjaminhuo Great, thank you!

**Comment by vajgi90** [comment]
We use Fluent-operator version 2.7.0, which uses Fluentd v.1.15.3. Unfortunately, we get the same error now: 
level=info msg="backoff timer done" actual=16.013265218s expected=16s
level=error msg="start Fluentd error" error="fork/exec /usr/local/bundle/bin/fluentd: no such file or directory"
level=info msg=backoff delay=32s
with the old image as well, which worked fine priorly. What can I do to get Fluentd to start?

**Comment by joshuabaird** [comment]
@vajgi90 It looks like the amd64 image on Dockerhub for `v1.15.3` has the bug.  We'll try to get this fixed.  Until then, you have two options:

* Use `ghcr.io/fluent/fluent-operator/fluentd:v1.15.3`
* Use `ghcr.io/fluent/fluent-operator/fluentd:v1.17.0`

**Comment by vajgi90** [comment]
Great, thank you so much for the quick response!
